### PR TITLE
discourse: update to 3.0.1, update plugins and mail-receiver to latest versions

### PR DESCRIPTION
**Pull request by pdg137**
###### Description of changes

This updates [discourse](https://github.com/discourse/discourse) to 3.0.1, the latest stable release.

I did it approximately as follows:

1. Add a line in update.py to fix gem calls that break across lines ([this line](https://github.com/discourse/discourse-prometheus/blob/main/plugin.rb#L15-L16) in discourse-prometheus was causing it to fail.)
2. Run the update scripts:
    ```
    ./pkgs/servers/web-apps/discourse/update.py update v3.0.1
    ./pkgs/servers/web-apps/discourse/update.py update-plugins
    ./pkgs/servers/web-apps/discourse/update.py update-mail-receiver
    ```
3. Edit `pkgs/servers/web-apps/discourse/rubyEnv/gemset.nix` to reference the platform-independent version of nokogiri and add the mini-portile2 dependency as suggested below. By default bundix selected the aarch64-linux hash, and the platform-independent gem did not build. 
4. Test with `nix-build -j 1 --keep-failed -A nixosTests.discourse`

The tests pass when this commit is applied to the release-22.11 branch but NOT on master.  For a working version see https://github.com/pdg137/nixpkgs/tree/discourse-3.0.1-22.11

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pdg137** [comment]
@Ma27 I just noticed you did some discourse updates recently.  Can you give me any pointers on getting this 3.0.1 upgrade cleaned up?

**Comment by wegank** [comment]
Please rebase against master, thanks!

File: pkgs/servers/web-apps/discourse/rubyEnv/gemset.nix [hunk 1]
```diff
@@ -1367,15 +1292,15 @@
     version = "2.5.8";
   };
   nokogiri = {
-    dependencies = ["mini_portile2" "racc"];
+    dependencies = ["racc"];
```
Comment (Ma27) [inline pkgs/servers/web-apps/discourse/rubyEnv/gemset.nix:1295]: @pdg137 the trick is to manually re-add `mini_portile2` to the `dependencies` section of `nokogiri`. No idea why this happens, but let's be honest: life's too short to debug rubygems issues :upside_down_face:

Comment (pdg137) [reply pkgs/servers/web-apps/discourse/rubyEnv/gemset.nix]: Thanks! That worked great. There's still also the problem of it getting the wrong hash, but that's easy enough to fix.

**Comment by pdg137** [comment]
I rebuilt this on master following @Ma27's suggestion.  Everything seems to run, but the tests do not pass, eventually generating messages like this forever:

```
discourse # curl: (35) OpenSSL/3.0.7: error:0200008A:rsa routines::invalid padding
discourse # curl: (35) OpenSSL/3.0.7: error:0200008A:rsa routines::invalid padding
discourse # curl: (35) OpenSSL/3.0.7: error:0200008A:rsa routines::invalid padding
discourse # curl: (35) OpenSSL/3.0.7: error:0200008A:rsa routines::invalid padding
discourse # curl: (35) OpenSSL/3.0.7: error:0200008A:rsa routines::invalid padding
```

It seems to work correctly when applied to the release-22.11 branch and I will be testing it more with that version.  I would be happy for more feedback!

**Comment by talyz** [comment]
Hi! Thanks for working on this! We generally don't package the stable versions of Discourse, but the beta versions, to better follow upstream behavior. I've opened a PR (#215523) that updates to the latest beta version instead and fixes a few other issues. I realize that we don't have any step-by-step guide on how to update Discourse, so I've added that as well. I took the liberty of copying your line split fix and assigned you co-author - if you're not okay with that, just tell me.

**Comment by pdg137** [comment]
Thanks @talyz! I am happy with my little fix getting merged in that way. FYI the 22.11 rebase of this has worked for me so far in some actual deployments, though I haven't tested everything yet.

I don't know what discourse intends exactly with their "beta" and stable releases but had the impression that only minor changes are supposed to get merged into nixos-22.11, so it seemed like getting it onto the stable release was the thing to do.  Do you expect the latest beta version to always go into 22.11 going forward?

**Comment by talyz** [comment]
Great!

Basically, they deploy their "beta" in production and "stable" is only meant to be used in environments with very specific needs. So yeah, their naming is pretty misleading. Version updates are allowed on the stable branch, as long as they're not breaking, so we can focus on updating the version in `master` and backporting it to stable if nothing breaks. If a discourse version has obvious breaking changes or changes that would require user action on upgrade, it should not go into stable and we would have to keep the versions in `master` and `release-xx.xx` separate. This hasn't happened yet, though, and I doubt it will be a frequent occurrence.

The `mini-portile2` issue only happened because of a bug (https://github.com/rubygems/rubygems/issues/6231) in the bundler version on the stable branch, which made the `update.py` script's workaround for a bug in bundix (https://github.com/nix-community/bundix/issues/88) ineffective. Basically, bundix gets confused by the different platform-specific packages and picks the wrong one. To work around this, the `update.py` script removes all platforms but the `ruby` platform, which is the one we want. However, due to the bundler bug, the lock file wasn't actually updated to remove the platforms. Since this is fixed in the bundler version on `master`, the workaround works as intended again.

**Pull request closed without merging**